File: backend/src/main/java/com/expensesage/service/PaymentService.java
package com.expensesage.service;

import java.util.List;

import com.expensesage.dto.PaymentCreateRequest;
import com.expensesage.model.Payment;
import com.expensesage.model.User;

public interface PaymentService {

    /**
     * Records a new payment made between users.
     *
     * @param request The DTO containing payment details.
     * @param payer   The user making the payment (authenticated user).
     * @return The newly created Payment entity.
     * @throws RuntimeException if recipient user not found, or group specified and
     *                          payer/recipient not in group.
     */
    Payment recordPayment(PaymentCreateRequest request, User payer);

    /**
     * Retrieves a payment by its ID, ensuring the current user was involved.
     *
     * @param paymentId   The ID of the payment.
     * @param currentUser The user requesting the details.
     * @return The Payment entity.
     * @throws RuntimeException if payment not found or user was not
     *                          payer/recipient.
     */
    Payment getPaymentById(Long paymentId, User currentUser);

    /**
     * Retrieves all payments where the current user was either the payer or the
     * recipient.
     *
     * @param currentUser The user whose payments to retrieve.
     * @return A list of Payment entities.
     */
    List<Payment> getPaymentsForUser(User currentUser);

    /**
     * Retrieves payments made between two specific users (in either direction).
     *
     * @param userA One user.
     * @param userB The other user.
     * @return A list of Payment entities between the two users.
     */
    List<Payment> getPaymentsBetweenUsers(User userA, User userB);

    /**
     * Retrieves payments related to a specific group.
     *
     * @param groupId     The ID of the group.
     * @param currentUser The user requesting the list (must be a member).
     * @return A list of Payment entities associated with the group.
     * @throws RuntimeException if group not found or user is not a member.
     */
    List<Payment> getPaymentsForGroup(Long groupId, User currentUser);

    /**
     * Deletes a payment, ensuring the current user was involved.
     *
     * @param paymentId   The ID of the payment to delete.
     * @param currentUser The user attempting the deletion.
     * @throws RuntimeException if payment not found or user was not involved.
     */
    void deletePayment(Long paymentId, User currentUser);

    /**
     * Updates an existing payment.
     * Only the original payer can update the payment.
     *
     * @param paymentId   The ID of the payment to update.
     * @param request     The DTO containing updated payment details.
     * @param currentUser The user attempting the update (must be the original payer).
     * @return The updated Payment entity.
     * @throws RuntimeException if payment not found, recipient not found, or permission denied.
     */
    Payment updatePayment(Long paymentId, PaymentCreateRequest request, User currentUser);

    // Deleting payments might be disallowed or restricted in a real system,
    // but could be added if needed.
    // void deletePayment(Long paymentId, User currentUser);
}

File: frontend/src/services/paymentService.ts
import apiClient from './api';
import { PaymentCreateRequest, PaymentResponseDto } from '../types/api';

// Records a new payment
export const recordPayment = async (paymentData: PaymentCreateRequest): Promise<PaymentResponseDto> => {
  try {
    const response = await apiClient.post<PaymentResponseDto>('/payments', paymentData);
    return response.data;
  } catch (error: any) {
    throw error.response?.data || new Error('Failed to record payment');
  }
};

// Deletes a specific payment
export const deletePayment = async (paymentId: number): Promise<void> => {
    try {
        await apiClient.delete(`/payments/${paymentId}`);
    } catch (error: any) {
        throw error.response?.data || new Error(`Failed to delete payment ${paymentId}`);
    }
};

// Updates an existing payment
export const updatePayment = async (paymentId: number, paymentData: PaymentCreateRequest): Promise<PaymentResponseDto> => {
    try {
        const formattedData = {
            ...paymentData,
            date: paymentData.date ? new Date(paymentData.date).toISOString().split('T')[0] : new Date().toISOString().split('T')[0],
        };
        const response = await apiClient.put<PaymentResponseDto>(`/payments/${paymentId}`, formattedData);
        return response.data;
    } catch (error: any) {
        throw error.response?.data || new Error(`Failed to update payment ${paymentId}`);
    }
};